### ISSUE TITLE
epoch searcher

### ISSUE THREAD

**Issue by danbryan** [DefiantLabs/cosmos-tax-cli#414]
I think it should be pretty easy to find all the epochs.  Being that only 1 happens per day, we can currently get 2 years of epoch history with ~ 730 calls, instead of 9million which is currently required.  

quick shell psuedo example
```
RPC_LATEST_KNOW_BLOCK=$(curl -s https://osmosisarchive-rpc.quickapi.com:443/status | jq .result.sync_info.latest_block_height)

N = $(osmosisd q epochs epoch-infos --node https://osmosisarchive-rpc.quickapi.com:443 -o json --height $RPC_LATEST_KNOW_BLOCK | jq -r .epochs | jq '.[] | select(.identifier == "day") | .current_epoch_start_height')

PREVIOUS = ((($current - 1)))

NMINUSONE = $(osmosisd q epochs epoch-infos --node https://osmosisarchive-rpc.quickapi.com:443 -o json --height $PREVIOUS | jq -r .epochs | jq '.[] | select(.identifier == "day") | .current_epoch_start_height')

```

TLDR  I think we should be able to easily find all epochs this way, and we should probably keep track of the epochs in the DB too.  

## Implementation Idea.

Create a task like we do for denom updates to find epoch events, and store them in the database.  This will probably only need to be ran once to populate the DB, we should assume whats in the DB is correct.  Then subsequent runs will happen every 24 hours and add the new days information.

When an indexer user whats to index the osmosis epochs they use dedicated flags.  This is separate from searching for random block events.

--plugin.osmosis-epoch-day-start 650
--plugin.osmosis-epoch-day-end 651

This will lookup in our database at which `current_epoch_start_height` did epoch 650, 651 each start at.

650
```
{
  "identifier": "day",
  "start_time": "2021-06-18T17:00:00Z",
  "duration": "86400s",
  "current_epoch": "650",
  "current_epoch_start_time": "2023-03-30T17:16:09.898160996Z",
  "epoch_counting_started": true,
  "current_epoch_start_height": "8945342"
}
```
651

```
{
  "identifier": "day",
  "start_time": "2021-06-18T17:00:00Z",
  "duration": "86400s",
  "current_epoch": "651",
  "current_epoch_start_time": "2023-03-31T17:16:09.898160996Z",
  "epoch_counting_started": true,
  "current_epoch_start_height": "8960478"
}
```
It will then search the block events at `8945342`, and `8960478`

**Comment by danbryan** [comment]
moving this to the backlog

**Comment by pharr117** [comment]
I have started working on this, see the following for more information on the direction its taking:

1. Feature writeup [here](https://github.com/DefiantLabs/cosmos-indexer/wiki/Osmosis-Epoch-and-Rewards-Indexing)
2. Initial push for Epoch Querying taking place in our Lens fork [here](https://github.com/DefiantLabs/lens/commit/4285c4f86664b500727f0b704917d8f3f2c26296)

**Comment by pharr117** [comment]
I have finished the first pass of this feature here: #434

It consists of:

1. Package update to pull in latest Lens fork version containing relevant updates
2. New model for storing Epochs as required by this feature
3. A new command run with `update-epochs` which:
    * Works for Osmosis currently, but attempts to be general enough in case this is needed elsewhere
    * Needs an Epoch identifier flag, defaults to `day`
    * Loops backwards from the current Osmosis Epoch until Epoch 1 (attempts to at least)
    * Can be rerun whenever to pick up the latest Epochs

Some findings during development made point (3) above impossible. Osmosis added the data attribute (the `current_epoch_start_height` value) that drives this command [here](https://github.com/osmosis-labs/osmosis/commit/44460ef087558ac7ab5dd8da7b165b198c3d0cfa), which was merged about ~170 days into the life of the Osmosis mainnet. The values for `current_epoch_start_height` stored in the history do not exist before then and default to "0" in the response.

The merge above ignores this problem and just reports an error and that it cannot continue when a "0" is found.

I have an idea for a fallback option that may work, using a method like the following:

1. Find all Blocks that fit this query: `"epoch_start.epoch_number EXISTS AND block.height < [last seen height]"`
2. Get the heights of all these blocks
3. Get Epoch Response for [each block, each block - 1]
4. Check which Epoch identifier changes in this time

A proof of concept for this is in the works right now, but I am also trying to push along Osmosis Incentives and Protorev Epoch indexing to not fall behind on that.

**Comment by danbryan** [comment]
Nice work, and yeah I would say the first 170 epochs are not a high priority item, more just a nice to have feature in the future.  If it's an easy lift and you want to do it before we do the osmosis re-index, no problem, but otherwise, i would say focus on the stuff thats most important to osmosis like protorev
